doc/tutorials/html_runner.py

```diff
@@ -0,0 +1,104 @@
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+import nbformat
+from nbconvert.preprocessors import ExecutePreprocessor
+import re
+import os
+import sys
+import argparse
+sys.path.append('@CMAKE_SOURCE_DIR@/testsuite/scripts')
+from importlib_wrapper import substitute_variable_values, mock_es_visualization
+
+parser = argparse.ArgumentParser(description='Process IPython notebooks.')
+parser.add_argument('--input', type=str,
+                    help='Path to the original IPython notebook')
+parser.add_argument('--output', type=str, nargs='?',
+                    help='Path to the processed IPython notebook')
+parser.add_argument('--substitutions', nargs='*',
+                    help='Variables to substitute')
+parser.add_argument('--scripts', nargs='*',
+                    help='Scripts to insert in new cells')
+args = parser.parse_args()
+
+
+def get_code_cells(nb):
+    return [c['source'] for c in nb['cells'] if c['cell_type'] == 'code']
+
+
+def set_code_cells(nb, new_cells):
+    i = 0
+    for c in nb['cells']:
+        if c['cell_type'] == 'code':
+            c['source'] = new_cells[i]
+            i += 1
+
+
+notebook_filepath = args.input
+notebook_filepath_edited = args.output or args.input + '~'
+notebook_dirname = os.path.dirname(notebook_filepath)
+new_values = args.substitutions or []
+new_cells = args.scripts or []
+
+# parse original notebook
+with open(notebook_filepath, encoding='utf-8') as f:
+    nb = nbformat.read(f, as_version=4)
+
+# add new cells containing the solutions
+for filepath in new_cells:
+    with open(filepath, encoding='utf-8') as f:
+        code = f.read()
+    # remove ESPResSo copyright header
+    m = re.search('# Copyright \(C\) \d+(?:\d+)? The ESPResSo project\n.+?'
+                  'If not, see <http://www\.gnu\.org/licenses/>\.\n', code, re.DOTALL)
+    if m and all(x.startswith('#') for x in m.group(0).strip().split('\n')):
+        code = re.sub('^(#\n)+', '', code.replace(m.group(0), ''), re.M)
+    # if matplotlib is imported in this script, split into two cells
+    if 'import matplotlib' in code:
+        cells_code = re.split('^((?:|.*\n)import matplotlib.*?)\n', code,
+                              maxsplit=1, flags=re.DOTALL)[1:]
+    else:
+        cells_code = [code]
+    # create new cells
+    cell_md = nbformat.v4.new_markdown_cell(source='Solution from ' + filepath)
+    nb['cells'].append(cell_md)
+    for cell_code in cells_code:
+        cell_code = nbformat.v4.new_code_cell(source=cell_code.strip())
+        nb['cells'].append(cell_code)
+
+# substitute global variables and disable OpenGL/Mayavi GUI
+unique_string = '\n##h7OZVDMsDxUPhqoVy2Mh0BTcoaFVbMUqEYCHYCDH\n'
```

Comment:
what is that string for?

Reply:
A comment line added between cells, helps splitting the source code back into notebook cells after variable substitution. It would be cleaner to use the uuid module instead of a hard-coded line.

Reply:
Fixed in 010d470